<a href="https://colab.research.google.com/github/Gustavens/Google-colab-AI/blob/main/Act_2_T4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Instalacion de paquetes

In [ ]:
!pip install NLTK

Bibliotecas a usar

In [ ]:
import nltk
import random
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn import datasets
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
#from sklearn.cross_validation import train_test_split

import matplotlib.pyplot as plt
pd.set_option('display.max_columns', None)
%matplotlib inline


**Descarga el corpus “wine dataset” dentro de scikit-learn y explóralo**

El conjunto de datos contiene los resultados de análisis químico de vinos cultivados en la misma región en Italia, pero derivados de tres cultivares diferentes, consta de 13 atributos y 3 tipos de vino.

In [ ]:
wine=load_wine()
print(wine)

{'data': array([[1.423e+01, 1.710e+00, 2.430e+00, ..., 1.040e+00, 3.920e+00,
        1.065e+03],
       [1.320e+01, 1.780e+00, 2.140e+00, ..., 1.050e+00, 3.400e+00,
        1.050e+03],
       [1.316e+01, 2.360e+00, 2.670e+00, ..., 1.030e+00, 3.170e+00,
        1.185e+03],
       ...,
       [1.327e+01, 4.280e+00, 2.260e+00, ..., 5.900e-01, 1.560e+00,
        8.350e+02],
       [1.317e+01, 2.590e+00, 2.370e+00, ..., 6.000e-01, 1.620e+00,
        8.400e+02],
       [1.413e+01, 4.100e+00, 2.740e+00, ..., 6.100e-01, 1.600e+00,
        5.600e+02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2,
 

In [ ]:
# Ver las características del dataset
print("\nKeys del dataset de vino:", wine.keys())

# Descripción del dataset (información sobre los atributos y clases)
print("\nDescripción del dataset de vino:\n", wine.DESCR)

# Nombres de los atributos (características)
print("\nNombres de los atributos (features):", wine.feature_names)

# Nombres de las clases (tipos de vino)
print("\nNombres de las clases (target names):", wine.target_names)

# Dimensiones de los datos (número de muestras, número de atributos)
print("\nDimensiones de los datos (data.shape):", wine.data.shape)

# Dimensiones de las etiquetas (target.shape)
print("Dimensiones de las etiquetas (target.shape):", wine.target.shape)

# Mostrar las primeras 5 filas de los datos (atributos)
print("\nPrimeras 5 filas de los datos (atributos):\n", wine.data[:5])

# Mostrar las primeras 5 etiquetas (clases)
print("\nPrimeras 5 etiquetas (clases):\n", wine.target[:5])

# Opcional: Convertir a un DataFrame de Pandas para una exploración más cómoda
df_wine = pd.DataFrame(wine.data, columns=wine.feature_names)
df_wine['wine_type'] = wine.target # Añadir las etiquetas como una columna
print("\nPrimeras 5 filas del DataFrame de Pandas:\n", df_wine.head())
print("\nConteo de tipos de vino:\n", df_wine['wine_type'].value_counts())

# 2. Separar las columnas en variables dependientes e independientes
print("\n--- Paso 2: Separando variables dependientes e independientes ---")
X = wine.data  # Variables independientes (atributos)
y = wine.target # Variable dependiente (etiquetas de clase)

print("Forma de X (atributos):", X.shape)
print("Forma de y (etiquetas):", y.shape)
# Descripción del dataset (información sobre los atributos y clases)
print("\nDescripción del dataset de vino:\n", wine.DESCR)

# Nombres de los atributos (características)
print("\nNombres de los atributos (features):", wine.feature_names)

# Nombres de las clases (tipos de vino)
print("\nNombres de las clases (target names):", wine.target_names)

# Dimensiones de los datos (número de muestras, número de atributos)
print("\nDimensiones de los datos (data.shape):", wine.data.shape)

# Dimensiones de las etiquetas (target.shape)
print("Dimensiones de las etiquetas (target.shape):", wine.target.shape)

# Mostrar las primeras 5 filas de los datos (atributos)
print("\nPrimeras 5 filas de los datos (atributos):\n", wine.data[:5])

# Mostrar las primeras 5 etiquetas (clases)
print("\nPrimeras 5 etiquetas (clases):\n", wine.target[:5])

# Opcional: Convertir a un DataFrame de Pandas para una exploración más cómoda
df_wine = pd.DataFrame(wine.data, columns=wine.feature_names)
df_wine['wine_type'] = wine.target # Añadir las etiquetas como una columna
print("\nPrimeras 5 filas del DataFrame de Pandas:\n", df_wine.head())
print("\nConteo de tipos de vino:\n", df_wine['wine_type'].value_counts())

# 2. Separar las columnas en variables dependientes e independientes
print("\n--- Paso 2: Separando variables dependientes e independientes ---")
X = wine.data  # Variables independientes (atributos)
y = wine.target # Variable dependiente (etiquetas de clase)

print("Forma de X (atributos):", X.shape)
print("Forma de y (etiquetas):", y.shape)




Keys del dataset de vino: dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names'])

Descripción del dataset de vino:
 .. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

:Number of Instances: 178
:Number of Attributes: 13 numeric, predictive attributes and the class
:Attribute Information:
    - Alcohol
    - Malic acid
    - Ash
    - Alcalinity of ash
    - Magnesium
    - Total phenols
    - Flavanoids
    - Nonflavanoid phenols
    - Proanthocyanins
    - Color intensity
    - Hue
    - OD280/OD315 of diluted wines
    - Proline
    - class:
        - class_0
        - class_1
        - class_2

:Summary Statistics:

============================= ==== ===== ======= =====
                                Min   Max   Mean     SD
============================= ==== ===== ======= =====
Alcohol:                      11.0  14.8    13.0   0.8
Malic Acid:                   0.74  5.80    2.34  1.12
Ash:                 

Utilizar el código para separar los datos en conjunto de entrenamiento y de prueba.

In [ ]:
# 3. Utilizar el código para separar los datos en conjunto de entrenamiento y de prueba
print("\n--- Paso 3: Separando datos en conjuntos de entrenamiento y prueba ---")
# Nota: train_test_split ha sido movido a sklearn.model_selection en versiones más recientes.
# La importación en el ejercicio era de sklearn.cross_validation, que está deprecada.
# He corregido la importación al inicio del script.

X_train, X_test, y_train, y_test = train_test_split(wine.data, wine.target, test_size=0.3, random_state=109)

print("Tamaño del conjunto de entrenamiento (X_train):", X_train.shape)
print("Tamaño del conjunto de prueba (X_test):", X_test.shape)
print("Tamaño de las etiquetas de entrenamiento (y_train):", y_train.shape)
print("Tamaño de las etiquetas de prueba (y_test):", y_test.shape)

print("\n¡Datos del vino cargados, explorados y divididos correctamente!")


--- Paso 3: Separando datos en conjuntos de entrenamiento y prueba ---
Tamaño del conjunto de entrenamiento (X_train): (124, 13)
Tamaño del conjunto de prueba (X_test): (54, 13)
Tamaño de las etiquetas de entrenamiento (y_train): (124,)
Tamaño de las etiquetas de prueba (y_test): (54,)

¡Datos del vino cargados, explorados y divididos correctamente!


En este paso clave del aprendizaje automático, trabajamos con un conjunto de datos sobre vinos y lo dividimos en dos partes con propósitos distintos:
Conjunto de Entrenamiento: Es el grupo de datos que utilizamos para enseñar al modelo. Básicamente, el algoritmo estudia los patrones y relaciones entre los 13 atributos de cada vino y su clasificación para poder predecir correctamente a futuro.

Conjunto de Prueba: Estos datos se dejan aparte y no se usan en el entrenamiento. Su función es evaluar el rendimiento del modelo una vez que ha aprendido, ayudándonos a medir qué tan bien puede generalizar sus predicciones a datos nuevos y desconocidos, simulando así una aplicación en el mundo real.

Esta separación es fundamental para asegurarnos de que el modelo no solo "memoriza" los datos de entrenamiento, sino que realmente aprende a reconocer patrones para hacer predicciones acertadas.


**Genera el modelo utilizando el conjunto de datos de entrenamiento.**

In [ ]:
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier # Importamos el clasificador de Árbol de Decisión

In [ ]:
# Cargar y preparar los datos (reutilizando el código anterior para asegurar la consistencia)
wine = load_wine()
X = wine.data
y = wine.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=109)

print("--- Paso 4: Generando el modelo de Árbol de Decisión con el conjunto de entrenamiento ---")

# 1. Generar el modelo utilizando el conjunto de datos de entrenamiento
# Creamos una instancia del clasificador de Árbol de Decisión
model = DecisionTreeClassifier(random_state=109) # Se usa random_state para reproducibilidad del árbol

# Entrenamos el modelo con los datos de entrenamiento
model.fit(X_train, y_train)

print("Modelo de Árbol de Decisión entrenado exitosamente.")

#El algoritmo de DecisionTreeClassifier ahora "aprende" las reglas y patrones de decisión a partir de estos datos para poder clasificar nuevos vinos.

--- Paso 4: Generando el modelo de Árbol de Decisión con el conjunto de entrenamiento ---
Modelo de Árbol de Decisión entrenado exitosamente.


In [ ]:
print("\n--- Paso 5: Realizando predicciones con el conjunto de datos de prueba ---")

# 2. Con el modelo construido, realiza predicciones utilizando el conjunto de datos de prueba
# y almacena los resultados en una variable nombrada y_pred.
y_pred = model.predict(X_test)

print("Predicciones realizadas con éxito.")
print(f"Primeras 10 predicciones: {y_pred[:10]}")
print(f"Primeras 10 etiquetas reales: {y_test[:10]}")

print("\nEl modelo ha sido entrenado y las predicciones han sido generadas.")


--- Paso 5: Realizando predicciones con el conjunto de datos de prueba ---
Predicciones realizadas con éxito.
Primeras 10 predicciones: [1 0 1 2 0 2 0 0 1 0]
Primeras 10 etiquetas reales: [0 0 1 2 0 1 0 1 1 0]

El modelo ha sido entrenado y las predicciones han sido generadas.


y_pred = model.predict(X_test):

Esta es la fase de predicción.
Una vez que el modelo está entrenado, usamos el método model.predict().

Le pasamos X_test (los atributos del conjunto de prueba, datos que el modelo nunca ha visto durante el entrenamiento).

El modelo utiliza las reglas que aprendió para predecir el tipo de vino para cada muestra en X_test.

Los resultados de estas predicciones se almacenan en la variable y_pred.

Evaluación del modelo.

In [ ]:
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics # Importamos el módulo metrics

# Cargar y preparar los datos (reutilizando el código anterior para asegurar la consistencia)
wine = load_wine()
X = wine.data
y = wine.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=109)

# Generar y entrenar el modelo de Árbol de Decisión
model = DecisionTreeClassifier(random_state=109)
model.fit(X_train, y_train)

# Realizar predicciones
y_pred = model.predict(X_test)

print("--- Evaluación inicial del modelo de Árbol de Decisión ---")
# Evaluar el modelo utilizando accuracy_score
accuracy = metrics.accuracy_score(y_test, y_pred)
print(f"Accuracy del modelo de Árbol de Decisión: {accuracy:.4f}")

# También podemos ver otras métricas para una evaluación más completa
print("\nOtras métricas de evaluación:")
print(f"Precision: {metrics.precision_score(y_test, y_pred, average='weighted'):.4f}")
print(f"Recall: {metrics.recall_score(y_test, y_pred, average='weighted'):.4f}")
print(f"F1-Score: {metrics.f1_score(y_test, y_pred, average='weighted'):.4f}")
print("\nMatriz de Confusión:\n", metrics.confusion_matrix(y_test, y_pred))
print("\nReporte de Clasificación:\n", metrics.classification_report(y_test, y_pred, target_names=wine.target_names))

--- Evaluación inicial del modelo de Árbol de Decisión ---
Accuracy del modelo de Árbol de Decisión: 0.8148

Otras métricas de evaluación:
Precision: 0.8156
Recall: 0.8148
F1-Score: 0.8095

Matriz de Confusión:
 [[19  2  0]
 [ 3 12  4]
 [ 0  1 13]]

Reporte de Clasificación:
               precision    recall  f1-score   support

     class_0       0.86      0.90      0.88        21
     class_1       0.80      0.63      0.71        19
     class_2       0.76      0.93      0.84        14

    accuracy                           0.81        54
   macro avg       0.81      0.82      0.81        54
weighted avg       0.82      0.81      0.81        54



Métricas Generales

Accuracy (Precisión global): 0.8148 → El modelo clasifica correctamente aproximadamente el 81.5% de los casos.

Precision promedio: 0.8156

Recall promedio: 0.8148

Promedio armónico entre la precisión (precision) y la sensiblidad (recall).
F1-Score promedio: 0.8095

La Clase 0 se predice con bastante precisión.

La Clase 1 tiene más confusiones, especialmente con la Clase 2.

La Clase 2 tiene un buen recall (0.93), lo que indica que la mayoría de los ejemplos reales de esta clase fueron correctamente identificados.

**Conclusión del resultado.**

El modelo tiene un rendimiento sólido general, especialmente en la Clase 0 y Clase 2. Sin embargo, podría beneficiarse de mejoras en la detección de la Clase 1.

**Mejoras a considerar**

Reducir la complejidad del Árbol de Decisión, ajustando parámetros como max_depth o min_samples_leaf, lo que ayuda a evitar el sobreajuste.

Usar GridSearchCV, una herramienta que busca automáticamente la mejor combinación de parámetros para optimizar el rendimiento del modelo.

Probar con modelos de ensamble, como RandomForestClassifier, que combinan varios árboles para lograr predicciones más robustas y precisas.

Al entrenar un modelo con DecisionTreeClassifier, hay varios parámetros que puedes ajustar para mejorar su rendimiento y evitar que se sobreentrene.

Uso de DecisionTreeClassifier

In [ ]:
# Imprimimos un encabezado para indicar que vamos a mejorar el modelo
print("\n--- Mejorando el modelo de Árbol de Decisión (Ajuste de Hiperparámetros) ---")

#  Modelo 1: Limitando la profundidad del árbol para evitar sobreajuste
# Un árbol muy profundo puede memorizar los datos de entrenamiento, lo que reduce su capacidad de generalizar
model_tuned_1 = DecisionTreeClassifier(max_depth=5, random_state=109)  # Limita la profundidad máxima a 5 niveles
model_tuned_1.fit(X_train, y_train)  # Entrenamos el modelo con los datos de entrenamiento
y_pred_tuned_1 = model_tuned_1.predict(X_test)  # Hacemos predicciones sobre los datos de prueba
accuracy_tuned_1 = metrics.accuracy_score(y_test, y_pred_tuned_1)  # Calculamos la precisión del modelo
print(f"Accuracy del Árbol de Decisión (max_depth=5): {accuracy_tuned_1:.4f}")

#  Modelo 2: Estableciendo un número mínimo de muestras por hoja
# Esto evita que el árbol cree nodos hoja con muy pocos datos, lo que también ayuda a reducir el sobreajuste
model_tuned_2 = DecisionTreeClassifier(min_samples_leaf=5, random_state=109)  # Requiere al menos 5 muestras por hoja
model_tuned_2.fit(X_train, y_train)
y_pred_tuned_2 = model_tuned_2.predict(X_test)
accuracy_tuned_2 = metrics.accuracy_score(y_test, y_pred_tuned_2)
print(f"Accuracy del Árbol de Decisión (min_samples_leaf=5): {accuracy_tuned_2:.4f}")

#  Modelo 3: Búsqueda automática de los mejores parámetros con GridSearchCV
# Esta técnica prueba múltiples combinaciones de parámetros para encontrar la que da mejor rendimiento
from sklearn.model_selection import GridSearchCV

# Definimos el espacio de búsqueda de parámetros
param_grid = {
    'max_depth': [3, 5, 7, 10],  # Profundidades posibles del árbol
    'min_samples_leaf': [1, 3, 5, 10],  # Tamaños mínimos de hojas
    'criterion': ['gini', 'entropy']  # Funciones para medir la calidad de las divisiones
}

# Configuramos GridSearchCV con validación cruzada de 5 pliegues
grid_search = GridSearchCV(DecisionTreeClassifier(random_state=109), param_grid, cv=5, scoring='accuracy')
grid_search.fit(X_train, y_train)  # Entrenamos el modelo con todas las combinaciones posibles



--- Mejorando el modelo de Árbol de Decisión (Ajuste de Hiperparámetros) ---
Accuracy del Árbol de Decisión (max_depth=5): 0.8148
Accuracy del Árbol de Decisión (min_samples_leaf=5): 0.8148


GridSearchCV(cv=5, estimator=DecisionTreeClassifier(random_state=109),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [3, 5, 7, 10],
                         'min_samples_leaf': [1, 3, 5, 10]},
             scoring='accuracy')

In [ ]:
# Mostramos los mejores parámetros encontrados
print(f"\nMejores parámetros encontrados por GridSearchCV: {grid_search.best_params_}")

# Evaluamos el mejor modelo encontrado
best_tree_model = grid_search.best_estimator_
y_pred_best_tree = best_tree_model.predict(X_test)
accuracy_best_tree = metrics.accuracy_score(y_test, y_pred_best_tree)
print(f"Accuracy del Árbol de Decisión optimizado con GridSearchCV: {accuracy_best_tree:.4f}")

# Mostramos un reporte detallado de clasificación
print("\nReporte de Clasificación del modelo optimizado:\n", metrics.classification_report(y_test, y_pred_best_tree, target_names=wine.target_names))



Mejores parámetros encontrados por GridSearchCV: {'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 1}
Accuracy del Árbol de Decisión optimizado con GridSearchCV: 0.8148

Reporte de Clasificación del modelo optimizado:
               precision    recall  f1-score   support

     class_0       0.86      0.90      0.88        21
     class_1       0.80      0.63      0.71        19
     class_2       0.76      0.93      0.84        14

    accuracy                           0.81        54
   macro avg       0.81      0.82      0.81        54
weighted avg       0.82      0.81      0.81        54



**Explorando Predicciones y Complementación de Texto con BERT y GPT**

In [ ]:
from transformers import pipeline
import os

# --- Modelos para español ---
# Para generar texto (similar a GPT): usaremos 'dccuchile/bert-base-spanish-wwm-cased' o 'dccuchile/bert-base-spanish-wwm-cased-finetuned-sqac'

# Inicializar pipelines
# Para generación de texto en español (GPT-like)

try:
    generator_es = pipeline('text-generation', model='dccuchile/gpt2-small-spanish')
except Exception as e:
    print(f"Error initializing Spanish text generation pipeline: {e}")
    print("Please ensure you are authenticated with Hugging Face Hub.")
    print("Run `huggingface-cli login` in your terminal or pass your token to the pipeline function.")
    # Decide how to handle the error if the pipeline cannot be initialized
    # For example, you might skip the Spanish generation part or use a different model.
    generator_es = None # Set to None to indicate failure


# Para rellenar huecos en español (BERT-like)
try:
    fill_mask_es = pipeline('fill-mask', model='dccuchile/bert-base-spanish-wwm-cased')
except Exception as e:
    print(f"Error initializing Spanish fill-mask pipeline: {e}")
    print("Please ensure you are authenticated with Hugging Face Hub.")
    fill_mask_es = None # Set to None

# --- Modelos para inglés ---
# Para generación de texto (GPT-like)
try:
    generator_en = pipeline('text-generation', model='gpt2') # GPT-2 base model
except Exception as e:
     print(f"Error initializing English text generation pipeline: {e}")
     generator_en = None # Set to None


# Para rellenar huecos (BERT-like)
try:
    fill_mask_en = pipeline('fill-mask', model='bert-base-uncased')
except Exception as e:
    print(f"Error initializing English fill-mask pipeline: {e}")
    fill_mask_en = None # Set to None


# --- Textos/Oraciones de prueba ---
textos_espanol = [
    "El perro ladró fuertemente y el gato se ",
    "La inteligencia artificial transformará la sociedad de maneras ",
    "El cambio climático es un desafío global que requiere ",
    "En la cima de la montaña, la vista era ",
    "El futuro de la tecnología es prometedor, pero también presenta "
]

textos_ingles = [
    "The dog barked loudly and the cat ",
    "Artificial intelligence will transform society in ways ",
    "Climate change is a global challenge that requires ",
    "At the top of the mountain, the view was ",
    "The future of technology is promising, but also presents "
]

print("--- Predicciones y Complementación (Español) ---")
if generator_es or fill_mask_es: # Only proceed if at least one Spanish model loaded
    for i, text in enumerate(textos_espanol):
        print(f"\n--- Texto {i+1} (Español) ---")
        print(f"Original: '{text}'")

        # Predicción con modelo generativo (GPT-like)
        if generator_es:
            print("\n[Generación GPT-like]:")
            try:
                # Ajustamos max_length para controlar la longitud de la generación
                # no_repeat_ngram_size para evitar repeticiones
                # num_return_sequences para obtener varias opciones
                generated_text_es = generator_es(text, max_length=len(text.split()) + 10, num_return_sequences=1, do_sample=True, temperature=0.7)
                for j, gen_seq in enumerate(generated_text_es):
                    print(f"  Opción {j+1}: '{gen_seq['generated_text']}'")
            except Exception as e:
                print(f"  Error during Spanish GPT-like generation: {e}")


        # Predicción con modelo de rellenar huecos (BERT-like)
        if fill_mask_es:
            text_masked = text + " [MASK]"
            print("\n[Rellenar Huecos BERT-like (palabra final)]: ")
            try:
                filled_mask_es = fill_mask_es(text_masked)
                for j, fill in enumerate(filled_mask_es[:3]): # Mostrar las top 3 predicciones
                    print(f"  Opción {j+1}: '{text_masked.replace('[MASK]', fill['token_str'])}' (score: {fill['score']:.4f})")
            except Exception as e:
                 print(f"  Error during Spanish BERT-like fill-mask: {e}")
else:
    print("\nSkipping Spanish predictions due to model loading errors.")


print("\n\n" + "="*50 + "\n") # Separador


print("--- Predicciones y Complementación (Inglés) ---")
if generator_en or fill_mask_en: # Only proceed if at least one English model loaded
    for i, text in enumerate(textos_ingles):
        print(f"\n--- Texto {i+1} (Inglés) ---")
        print(f"Original: '{text}'")

        # Predicción con modelo generativo (GPT-like)
        if generator_en:
            print("\n[Generación GPT-like]:")
            try:
                generated_text_en = generator_en(text, max_length=len(text.split()) + 10, num_return_sequences=1, do_sample=True, temperature=0.7)
                for j, gen_seq in enumerate(generated_text_en):
                    print(f"  Opción {j+1}: '{gen_seq['generated_text']}'")
            except Exception as e:
                print(f"  Error during English GPT-like generation: {e}")

        # Predicción con modelo de rellenar huecos (BERT-like)
        if fill_mask_en:
            text_masked = text + " [MASK]"
            print("\n[Rellenar Huecos BERT-like (palabra final)]: ")
            try:
                filled_mask_en = fill_mask_en(text_masked)
                for j, fill in enumerate(filled_mask_en[:3]): # Mostrar las top 3 predicciones
                    print(f"  Opción {j+1}: '{text_masked.replace('[MASK]', fill['token_str'])}' (score: {fill['score']:.4f})")
            except Exception as e:
                print(f"  Error during English BERT-like fill-mask: {e}")
else:
     print("\nSkipping English predictions due to model loading errors.")

Error initializing Spanish text generation pipeline: dccuchile/gpt2-small-spanish is not a local folder and is not a valid model identifier listed on 'https://huggingface.co/models'
If this is a private repository, make sure to pass a token having permission to this repo either by logging in with `huggingface-cli login` or by passing `token=<your_token>`
Please ensure you are authenticated with Hugging Face Hub.
Run `huggingface-cli login` in your terminal or pass your token to the pipeline function.


Device set to use cpu
Device set to use cpu
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Device set to use cpu


--- Predicciones y Complementación (Español) ---

--- Texto 1 (Español) ---
Original: 'El perro ladró fuertemente y el gato se '

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'El perro ladró fuertemente y el gato se  fue' (score: 0.3598)
  Opción 2: 'El perro ladró fuertemente y el gato se  escapó' (score: 0.2374)
  Opción 3: 'El perro ladró fuertemente y el gato se  marchó' (score: 0.0731)

--- Texto 2 (Español) ---
Original: 'La inteligencia artificial transformará la sociedad de maneras '

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'La inteligencia artificial transformará la sociedad de maneras  diferentes' (score: 0.4252)
  Opción 2: 'La inteligencia artificial transformará la sociedad de maneras  :' (score: 0.1431)
  Opción 3: 'La inteligencia artificial transformará la sociedad de maneras  distintas' (score: 0.1379)

--- Texto 3 (Español) ---
Original: 'El cambio climático es un desafío global que requiere '

[Rellenar Huecos BERT-like (palabra final)]

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


  Opción 1: 'En la cima de la montaña, la vista era  :' (score: 0.1968)
  Opción 2: 'En la cima de la montaña, la vista era  impresionante' (score: 0.1014)
  Opción 3: 'En la cima de la montaña, la vista era  espectacular' (score: 0.0715)

--- Texto 5 (Español) ---
Original: 'El futuro de la tecnología es prometedor, pero también presenta '

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'El futuro de la tecnología es prometedor, pero también presenta  problemas' (score: 0.2874)
  Opción 2: 'El futuro de la tecnología es prometedor, pero también presenta  desafíos' (score: 0.1582)
  Opción 3: 'El futuro de la tecnología es prometedor, pero también presenta  :' (score: 0.1218)



--- Predicciones y Complementación (Inglés) ---

--- Texto 1 (Inglés) ---
Original: 'The dog barked loudly and the cat '

[Generación GPT-like]:


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


  Opción 1: 'The dog barked loudly and the cat iced at him. He took her back'

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'The dog barked loudly and the cat  .' (score: 0.6324)
  Opción 2: 'The dog barked loudly and the cat  ?' (score: 0.1686)
  Opción 3: 'The dog barked loudly and the cat  !' (score: 0.1430)

--- Texto 2 (Inglés) ---
Original: 'Artificial intelligence will transform society in ways '

[Generación GPT-like]:


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


  Opción 1: 'Artificial intelligence will transform society in ways  that we won't be able to'

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'Artificial intelligence will transform society in ways  .' (score: 0.9627)
  Opción 2: 'Artificial intelligence will transform society in ways  ;' (score: 0.0361)
  Opción 3: 'Artificial intelligence will transform society in ways  !' (score: 0.0006)

--- Texto 3 (Inglés) ---
Original: 'Climate change is a global challenge that requires '

[Generación GPT-like]:


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


  Opción 1: 'Climate change is a global challenge that requires  a sustained effort and investment to address it'

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'Climate change is a global challenge that requires  .' (score: 0.7969)
  Opción 2: 'Climate change is a global challenge that requires  ;' (score: 0.1859)
  Opción 3: 'Climate change is a global challenge that requires  !' (score: 0.0076)

--- Texto 4 (Inglés) ---
Original: 'At the top of the mountain, the view was '

[Generación GPT-like]:


Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


  Opción 1: 'At the top of the mountain, the view was  roughly 300 meters above the ground'

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'At the top of the mountain, the view was  .' (score: 0.8131)
  Opción 2: 'At the top of the mountain, the view was  ;' (score: 0.0979)
  Opción 3: 'At the top of the mountain, the view was  !' (score: 0.0287)

--- Texto 5 (Inglés) ---
Original: 'The future of technology is promising, but also presents '

[Generación GPT-like]:
  Opción 1: 'The future of technology is promising, but also presents ills and challenges, and there is no'

[Rellenar Huecos BERT-like (palabra final)]: 
  Opción 1: 'The future of technology is promising, but also presents  .' (score: 0.9374)
  Opción 2: 'The future of technology is promising, but also presents  ;' (score: 0.0595)
  Opción 3: 'The future of technology is promising, but also presents  !' (score: 0.0015)
